// @cowprotocol/ts-dune-client/dist/queryParameter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryParameter = exports.ParameterType = void 0;
var ParameterType;
(function (ParameterType) {
    ParameterType["TEXT"] = "text";
    ParameterType["NUMBER"] = "number";
    ParameterType["DATE"] = "date";
    ParameterType["ENUM"] = "enum";
})(ParameterType = exports.ParameterType || (exports.ParameterType = {}));
class QueryParameter {
    type;
    value;
    name;
    constructor(type, name, value) {
        this.type = type;
        this.value = value.toString();
        this.name = name;
    }
    static text(name, value) {
        return new QueryParameter(ParameterType.TEXT, name, value);
    }
    static number(name, value) {
        // TODO - investigate large numbers here...
        return new QueryParameter(ParameterType.NUMBER, name, value.toString());
    }
    static date(name, value) {
        return new QueryParameter(ParameterType.DATE, name, value.toString());
    }
    static enum(name, value) {
        return new QueryParameter(ParameterType.ENUM, name, value.toString());
    }
}
exports.QueryParameter = QueryParameter;
